accepted/0000-validate-signatures.md

```diff
@@ -0,0 +1,509 @@
+ and at the `registry.npmjs.org/.well-known/npm-signature-keys` URL
+
+The `verify-signatures` should check that the version created time is
+within the `expires` time set on the public key.
+
+### Supporting third-party npm registries signing packages
+
+Our long-term goal for all software to be signed and verified in a transparent
+way.
+
+We want to make it possible for third-party npm registries (e.g. GitHub
+Packages, Artifactory, Verdaccio etc) to start signing published packages and
+allow the CLI to verify these signatures with minimal user intervention.
+
+#### Setting up package signing on a third-party npm registry:
+
+- Set up a secure signing key, preferrably using HSM-backed key in a cloud KMS
+(e.g. AWS KMS, Azure Vault, GCP KMS etc)
+- Add a new API endpoint to the registry host:
+  `https://registry.host/.well-known/npm-signature-keys`
+- Configure the endpoint to return the public key and key metadata (see payload
+  example under "Fetching the public key")
+- Start signing package versions on publish, adding the signature (`sig`) and
+  `keyid` to the version packument, nested in the `dist.signatures` array (see
+  example under "Packument example with new `signatures` field")
+
+#### Verifying third-party registry signatures using the npm CLI
+
+There's currently no reliable (or is there?) way to differentiate a mirror/proxy
+from a originating third-party npm registry that hosts packages not on
+npmjs.org.
+
+Users will be asked to manually approve unknown keys when verifying signatures
+that reference untrusted `keyid`s. This is similar to how ssh works when
+accessing a new server. Trusting new keys will add the keys to the `.npmrc`
+file, which should be committed to source.
+
+**Attack scenario: malicious npm mirror resigns packages and tries to trick users
+into trusting it's keys**:
+
+If a user has configured to use a proxy/mirror for npm packages, and set the
+global registry, for example:
+
+`.npmrc`
+```
+registry=https://fast-npm-mirror.tld
+```
+
+This registry could be resigning packages it serves and serve it's own public
+signing keys at: https://fast-npm-mirror.tld/.well-known/npm-signature-keys
+
+We want to make sure the response hasn't been tampered with (MITM attack) before
+trusting the new signing keys.
+
+The CLI will prompt a user to manually trust new signing keys when verifying a
+package that has a signature and untrusted `keyid`.
+
+```
+$ npm verify-signatures
+project@1.0.0 $HOME/work/project
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!                  WARNING: FOUND UNTRUSTED SIGNATURE KEYS                   !!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Found untrusted signature keys at https://fast-npm-mirror.tld/.well-known/npm-signature-keys
+
+Are you sure you trust this host? (yes/no)
+> no
+```
+
+Trusted keys are added to the `.npmrc`:
+
+```
+[public-signature-key]
```

Comment:
can you define more than one trusted key using this notation in the `npmrc`? (I've never written one).

Reply:
(Is that even desirable? Maybe not.)

Reply:
Oh this syntax isn't correct, it should be `public-signature-key[]=`, but maybe worth naming it something like `trusted-signature-key[]=` would be clearer what this is 🤔  Also thinking we might not want to include all the attributes here as you might want to be able to update expiry on the key to rotate it etc. Maybe just `keyid` and `registry`. You'll want to be able to set multiple keys, for different registries.

This syntax is similar to how you can configure `ca`'s https://docs.npmjs.com/cli/v8/using-npm/config#ca

Reply:
@feelepxyz `ini` does support Objects combining to form Arrays although if we want nested value setting it does look like the syntax has to change a bit.

Here's an example of a valid/supported syntax we could use:

```ini
[public-signature-key.SHA256:{{SHA256_PUBLIC_KEY}}]
expires=2023-12-17T23:57:40-05:00
keytype=ecdsa-sha2-nistp256
sigtype=ecdsa-sha2-nistp256
key={{B64_PUBLIC_KEY}}
```

In this case, you'd have to specify a unique key id (but that's already a requirement of the spec) & then the keys would be grouped into an Array of values/config properly.

Feel free to try this out:
```js
const ini = require("ini")
let text = `
[public-signature-key.SHA256:123]
expires=2023-12-17T23:57:40-05:00
keytype=ecdsa-sha2-nistp256
sigtype=ecdsa-sha2-nistp256
key=123

[public-signature-key.SHA256:456]
expires=2023-12-17T23:57:40-05:00
keytype=ecdsa-sha2-nistp256
sigtype=ecdsa-sha2-nistp256
key=456

[public-signature-key.SHA256:789]
expires=2023-12-17T23:57:40-05:00
keytype=ecdsa-sha2-nistp256
sigtype=ecdsa-sha2-nistp256
key=789
`
console.log(ini.parse(text))
```

Reply:
@darcyclarke thanks for pointing this out! I think is what we want 👍  I see that it gets nested in the object as expected when running `npm config list --json` 👌 :

```
"public-signature-key": {
  "SHA256:123": {
    "expires": "2023-12-17T23:57:40-05:00",
    "keytype": "ecdsa-sha2-nistp256",
    "sigtype": "ecdsa-sha2-nistp256",
    "key": "123"
  },
  "SHA256:456": {
    "expires": "2023-12-17T23:57:40-05:00",
    "keytype": "ecdsa-sha2-nistp256",
    "sigtype": "ecdsa-sha2-nistp256",
    "key": "456"
  }
}
```